realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java

```diff
@@ -465,4 +479,78 @@ public void onChange(RealmResults<Dog> collection, @Nullable OrderedCollectionCh
             }
         });
     }
+
+    // To reproduce https://github.com/realm/realm-java/issues/5507
+    // 1. Add listener to a collection
+    // A. change the collection in a background thread
+    // 2. Remove the listener
+    // 3. Add another listener
+    // 4. the listener added in step 3 should be triggered with change set in step A
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_bug5507() throws InterruptedException {
+        // FIXME: See https://github.com/realm/realm-object-store/issues/605
+        if (type == ObservablesType.REALM_RESULTS) looperThread.testComplete();
```

Comment:
I think you want to return here as well. As the rest of the code still runs I think.